### ISSUE TITLE
First-class mixins

### ISSUE THREAD

**Issue by K4rakara** [sass/sass#3439]
I'm implementing mixins for a component library that will have multiple implementations for different frameworks. In doing this, I've found that in order to allow the various implementations to define how the various states of the components are selected, I would greatly benefit from first-class mixins, which don't currently exist in sass. Consider the following:

```scss
@use "sass:string";

@function string-has-prefix($string, $prefix) {
  @return string.slice($string, 1, string.length($prefix)) == $prefix;
}

@function string-has-suffix($string, $suffix) {
  @return string.slice($string, (string.length($suffix) * -1), -1) == $suffix;
}

@function string-remove-prefix($string, $prefix) {
  @if has-prefix($string, $prefix) {
    @return string.slice($string, string.length($prefix) + 1);
  } @else {
    @return $string;
  }
}

@mixin my-mixin($some-dynamic-property) {
  // Allow statically defining the property
  @if $some-dynamic-property == variant1 {
    display: block; // For example
  } @else if $some-dynamic-property == variant2 {
    display: inline-block; 
  // Allow specifying to have the mixin define a BEM "modifier" selector.
  } @else if $some-dynamic-property == modifier {
    &--variant-1 { display: block; }
    &--variant-2 { display: inline-block; }
  // Allow specifying a string formatted like attr(foo) to have the mixin set the property based
  // on what an attribute is set to on the target (&).
  } @else if (
    string-has-prefix($some-dynamic-property, "attr(")
    && string-has-suffix($some-dynamic-property, ")")
  ) {
    $some-dynamic-property: string-remove-prefix($some-dynamic-property, "attr(");
    $some-dynamic-property: string-remove-suffix($some-dynamic-property, ")");
    
    &[#{$some-dynamic-property}=variant1] { display: block; }
    &[#{$some-dynamic-property}=variant2] { display: inline-block; }
  } @else {
    @error "...";
  }
}

.foo {
    background-color: red;
    @include my-mixin(variant1);
}

.bar {
    background-color: green;
    @include my-mixin(variant2);
}

.baz {
    background-color: blue;
    @include my-mixin(modifier);
}

.qux {
    background-color: magenta;
    @include my-mixin(attr(data-dynamic-property));
}

/*
.foo {
  background-color: red;
  display: block;
}

.bar {
  background-color: green;
  display: inline-block;
}

.baz {
  background-color: blue;
}
.baz--variant-1 {
  display: block;
}
.baz--variant-2 {
  display: inline-block;
}

.qux {
  background-color: magenta;
}
.qux[data-dynamic-property=variant1] {
  display: block;
}
.qux[data-dynamic-property=variant2] {
  display: inline-block;
}
*/
```

This works for one mixin, but if there are a lot of mixins with properties specified this way, the code balloons quickly. How can we make this better? First class mixins!

Here is the same mixin, with a helper that would be enabled by first-class mixins:

```scss
@use "sass:meta";
@use "sass:string";

@function string-has-prefix($string, $prefix) {
  @return string.slice($string, 1, string.length($prefix)) == $prefix;
}

@function string-has-suffix($string, $suffix) {
  @return string.slice($string, (string.length($suffix) * -1), -1) == $suffix;
}

@function string-remove-prefix($string, $prefix) {
  @if has-prefix($string, $prefix) {
    @return string.slice($string, string.length($prefix) + 1);
  } @else {
    @return $string;
  }
}

@mixin noop-transform($variant) { }

@mixin dynamic-property($value, $variants: (), $transform: meta.get-mixin(noop-transform)) {
  // Use a loop to work around sasses lack of a way to early-exit mixins...
  // And if statements to work around sasses lack of a @break and @continue...
  $continue-while: true;
  $strategy: variants;
  @while ($continue-while) {
    @if $strategy == variants {
      $continue-each: true;
      @each $variant in $variants {
        @if $continue-each && $value == $variant {
          @include meta.apply($transform, $variant);
          $continue-each: false;
          $continue-while: false;
        }
      }
      $strategy: modifier;
    } @else if $strategy == modifier {
      @if $value == modifier {
        @each $variant in $variants {
          &--#{$variant} {
            @include meta.apply($transform, $variant);
          }
        }
        $continue-while: false;
      }
      $strategy: attribute;
    } @else if $strategy == attribute {
      @if (
        string-has-prefix($value, "attr(")
        && string-has-suffix($value, ")")
      ) {
        $value: string-remove-prefix($value, "attr(");
        $value: string-remove-suffix($value, ")");
        @each $variant in $variants {
          &[#{$value}=#{$variant}] {
            @include meta.apply($transform, $variant);
          }
        }
        $continue-while: false;
      }
      $strategy: null;
    } @else {
      @error "...";
    }
  }
}

@mixin -my-mixin-transform($variant) {
  @if $variant == variant1 {
    display: block;
    // We don't do it here, but you could emit as many styles as you want here. 
  } @else if $variant == variant2 {
    display: inline-block;
  } @else {
    @error "...";
  }
}

@mixin my-mixin($some-dynamic-property) {
  @include dynamic-property($some-dynamic-property,
    $variants: (variant1, variant2),
    $transform: meta.get-mixin(-my-mixin-transform));
}
```

In this case, the helper implementation, is much bigger than inline implementation, but if you have numerous mixins with these "dynamic properties", the code becomes much more maintainable with the helper.

# Detailed Specification

Firstly, for mixins to be a "first-class" value, a new value type must be added -- `mixin`: 
```scss
@use "sass:meta";

@mixin foo { }

$my-mixin: meta.get-mixin(foo);

@debug meta.type-of($my-mixin); // mixin
```

Secondly, as seen in the previous example, a new function would need to be added to the `sass:meta` module -- `get-mixin`.
This function would act almost identically to `get-function`, but with mixins instead of functions. The only difference would be that it would not take a `$css` keyword argument to specify css fallback, since mixins don't exist in css.

Finally, a new mixin would need to be added to the `sass:meta` module - `apply`.
This mixin would simply pass everything down the the specified mixin:
```scss
@use "sass:meta";

@mixin foo($a, $b, $c) {
  @debug $a; // 1
  @debug $b; // 2
  @debug $c; // 3
}

.bar {
  @include meta.apply(meta.get-mixin(foo), 1, 2, 3);
}
```

And to wrap things up, to pass a content block to a mixin dynamically, all you need to do is this, since `meta.apply` is a mixin:
```scss
@use "sass:meta";

@mixin verbatim {
  @content;
}

.foo {
  @include meta.apply(meta.get-mixin(verbatim)) {
    background-color: red;
  }
}

/*
.foo {
  background-color: red;
}
*/
```

# TL;DR

- New mixin in `sass:meta` -- `apply`:
  `meta.apply($mixin, $args...) <content block>`
  When `@include`d, this mixin applies the specified `$mixin` with the provided `$args`, and optionally, the provided content block.
  The `$mixin` should be a `mixin` returned by `meta.get-mixin()`.
  
- New function in `sass:meta` -- `get-mixin`:
  `meta.get-mixin($name, $module: null)`
  Returns the mixin named `$name`.
  If `$module` is null, this returns the mixin named `$name` without a namespace. Otherwise, `$module` must be a string matching the namespace of a `@use` rule in the current file, in which case this returns the mixin in that module named `$name`.
  The returned mixin can be applied using `meta.apply()`.
  
- New value type: `mixin`.

**Comment by K4rakara** [comment]
I tried implementing this in dart sass, but hit a roadblock: `dart run sass`, `dart compile jit-snapshot`, and `dart compile exe bin/sass.dart && bin/sass.exe` don't seem to reflect my changes to the code when ran. Maybe I'm doing something dumb, I haven't used dart before, but I'm fairly certain my code should work if I could get the tooling to.

**Comment by nex3** [comment]
Duplicate of #626
